File: src/common/main/java/hauveli/fishcasting/hexcasting/recipe/brainsweep/FishingVoidEntityTypeIngredient.java
package hauveli.fishcasting.hexcasting.recipe.brainsweep;

import at.petrak.hexcasting.common.recipe.ingredient.brainsweep.BrainsweepeeIngredientType;
import at.petrak.hexcasting.common.recipe.ingredient.brainsweep.EntityTypeIngredient;
import com.li64.tide.Tide;
import com.li64.tide.TideConfig;
import com.mojang.serialization.Codec;
import com.mojang.serialization.MapCodec;
import com.mojang.serialization.codecs.RecordCodecBuilder;
import net.minecraft.core.registries.BuiltInRegistries;
import net.minecraft.core.registries.Registries;
import net.minecraft.network.RegistryFriendlyByteBuf;
import net.minecraft.network.chat.Component;
import net.minecraft.network.codec.ByteBufCodecs;
import net.minecraft.network.codec.StreamCodec;
import net.minecraft.server.level.ServerLevel;
import net.minecraft.world.entity.Entity;
import net.minecraft.world.entity.EntityType;
import net.minecraft.world.level.Level;

import java.util.Objects;
import java.util.Optional;

public class FishingVoidEntityTypeIngredient extends EntityTypeIngredient {
    //@Override
    public Entity exampleEntity(Level level) {
        return this.entityType.create(level);
    }

    public FishingVoidEntityTypeIngredient(EntityType<?> entityType) {
        super(entityType);
    }

    public boolean isFishableDimension(Entity entity) {
        Level level = entity.level();

        return Tide.CONFIG.general.fishableVoidHeights.stream().anyMatch(entry ->
                Objects.equals(entry.dimension, level.dimension().location().toString()));
    }

    private int getHeightFromTideConfig(Entity entity) {
        // check dimension too
        Optional<TideConfig.General.VoidHeightEntry> entry = Tide.CONFIG.general.fishableVoidHeights.stream()
                .filter(e -> e.dimension.equals(entity.level().dimension().location().toString()))
                .findFirst();

        if (entry.isPresent()) {

            int height = entry.get().height;
            TideConfig.General.VoidHeightEntry.Type type = entry.get().type;
            // to consider: fishing above the world height? I think that would also be cool...
            switch (type) {
                case TideConfig.General.VoidHeightEntry.Type.ABSOLUTE -> {
                    return height;
                }
                case TideConfig.General.VoidHeightEntry.Type.RELATIVE_TO_BOTTOM -> {
                    return entity.level().getMinBuildHeight() + height;
                }
                case TideConfig.General.VoidHeightEntry.Type.RELATIVE_TO_TOP -> {
                    return entity.level().getMaxBuildHeight() + height;
                }
            }
        }
        return Integer.MIN_VALUE;
    }

    private boolean inVoidFluid(Entity entity) {
        double entityY = entity.position().y;
        return entityY <= getHeightFromTideConfig(entity);
    }

    @Override
    public boolean test(Entity entity, ServerLevel level) {
        // entity types are singletons
        return entity.getType() == this.entityType
                && this.inVoidFluid(entity);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        FishingVoidEntityTypeIngredient that = (FishingVoidEntityTypeIngredient) o;
        return Objects.equals(entityType, that.entityType);
    }

    public static class Type implements BrainsweepeeIngredientType<FishingVoidEntityTypeIngredient> {
        public static final MapCodec<FishingVoidEntityTypeIngredient> CODEC = RecordCodecBuilder.mapCodec(instance -> instance.group(
                BuiltInRegistries.ENTITY_TYPE.byNameCodec().fieldOf("entityType").forGetter(FishingVoidEntityTypeIngredient::getEntityType)
        ).apply(instance, FishingVoidEntityTypeIngredient::new));
        public static final StreamCodec<RegistryFriendlyByteBuf, FishingVoidEntityTypeIngredient> STREAM_CODEC = StreamCodec.composite(
                ByteBufCodecs.registry(Registries.ENTITY_TYPE), FishingVoidEntityTypeIngredient::getEntityType,
                FishingVoidEntityTypeIngredient::new
        );

        @Override
        public MapCodec<FishingVoidEntityTypeIngredient> codec() {
            return CODEC;
        }

        @Override
        public StreamCodec<RegistryFriendlyByteBuf, FishingVoidEntityTypeIngredient> streamCodec() {
            return STREAM_CODEC;
        }
    }
}

File: src/common/main/kotlin/hauveli/fishcasting/casting/recipe/brainsweep/FishingVoidEntityTypeIngredient.kt
package hauveli.fishcasting.casting.recipe.brainsweep

import at.petrak.hexcasting.common.recipe.ingredient.brainsweep.BrainsweepeeIngredientType
import at.petrak.hexcasting.common.recipe.ingredient.brainsweep.EntityTypeIngredient
import com.li64.tide.Tide
import com.li64.tide.TideConfig
import com.mojang.serialization.MapCodec
import com.mojang.serialization.codecs.RecordCodecBuilder
import net.minecraft.core.registries.BuiltInRegistries
import net.minecraft.core.registries.Registries
import net.minecraft.network.RegistryFriendlyByteBuf
import net.minecraft.network.codec.ByteBufCodecs
import net.minecraft.network.codec.StreamCodec
import net.minecraft.server.level.ServerLevel
import net.minecraft.world.entity.Entity
import net.minecraft.world.entity.EntityType
import net.minecraft.world.level.Level
import java.util.function.Function


class FishingVoidEntityTypeIngredient(
    entityType: EntityType<*>?) : EntityTypeIngredient(entityType) {
    fun isFishableDimension(entity: Entity): Boolean {
        val level = entity.level()

        return Tide.CONFIG.general.fishableVoidHeights.stream().anyMatch { entry: TideConfig.General.VoidHeightEntry? ->
            entry!!.dimension == level.dimension().location().toString()
        }
    }

    private fun getHeightFromTideConfig(entity: Entity): Int {
        // check dimension too, void height may vary
        val entry = Tide.CONFIG.general.fishableVoidHeights.stream()
            .filter { e: TideConfig.General.VoidHeightEntry? ->
                e!!.dimension == entity.level().dimension().location().toString()
            }
            .findFirst()

        if (entry.isPresent) {
            val height = entry.get().height
            val type = entry.get().type
            // to consider: fishing above the world height? I think that would also be cool...
            return when (type) {
                TideConfig.General.VoidHeightEntry.Type.ABSOLUTE -> {
                    height
                }

                TideConfig.General.VoidHeightEntry.Type.RELATIVE_TO_BOTTOM -> {
                    entity.level().minBuildHeight + height
                }

                TideConfig.General.VoidHeightEntry.Type.RELATIVE_TO_TOP -> {
                    entity.level().maxBuildHeight + height
                }
            }
        }
        return Int.MIN_VALUE
    }

    private fun inVoidFluid(entity: Entity): Boolean {
        val entityY = entity.position().y
        return entityY <= getHeightFromTideConfig(entity)
    }

    override fun test(entity: Entity, level: ServerLevel?): Boolean {
        return entity.type === this.entityType
                && this.inVoidFluid(entity)
    }

    override fun equals(o: Any?): Boolean {
        if (this === o) return true
        if (o == null || javaClass != o.javaClass) return false
        val that = o as FishingVoidEntityTypeIngredient
        return entityType == that.entityType
    }

    class Type : BrainsweepeeIngredientType<FishingVoidEntityTypeIngredient?> {
        override fun codec(): MapCodec<FishingVoidEntityTypeIngredient?>? {
            return CODEC
        }

        override fun streamCodec(): StreamCodec<RegistryFriendlyByteBuf?, FishingVoidEntityTypeIngredient?> {
            return STREAM_CODEC
        }

        companion object {
            val CODEC: MapCodec<FishingVoidEntityTypeIngredient?>? =
                RecordCodecBuilder.mapCodec<FishingVoidEntityTypeIngredient?>(
                    Function { instance: RecordCodecBuilder.Instance<FishingVoidEntityTypeIngredient?>? ->
                        instance!!.group(
                            BuiltInRegistries.ENTITY_TYPE.byNameCodec().fieldOf("entityType")
                                .forGetter<FishingVoidEntityTypeIngredient?>(
                                    { obj: FishingVoidEntityTypeIngredient? -> obj!!.getEntityType() })
                        ).apply<FishingVoidEntityTypeIngredient?>(
                            instance,
                            { entityType: EntityType<*>? -> FishingVoidEntityTypeIngredient(entityType) })
                    })
            val STREAM_CODEC: StreamCodec<RegistryFriendlyByteBuf?, FishingVoidEntityTypeIngredient?> =
                StreamCodec.composite<RegistryFriendlyByteBuf?, FishingVoidEntityTypeIngredient?, EntityType<*>?>(
                    ByteBufCodecs.registry(Registries.ENTITY_TYPE),
                    { obj: FishingVoidEntityTypeIngredient? -> obj!!.getEntityType() },
                    { entityType: EntityType<*>? -> FishingVoidEntityTypeIngredient(entityType) }
                )
        }
    }
}